### PR TITLE
ondsel: init at 2024.1.0

### PR DESCRIPTION
**Pull request by b-eyselein**
## Description of changes

This pull request adds [Ondsel](https://ondsel.com/), a FreeCAD fork with a modernized UI.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by b-eyselein** [comment]
If this package needs a maintainer to be merged, I could do that.

File: pkgs/by-name/on/ondsel/package.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        --replace 'Exec=AppRun' 'Exec=${pname}'
+      cp -r ${contents}/usr/share/icons $out/share
+    '';
+
+
+  meta = with lib; {
+    description = "An engineering suite with an open-source heart (Solo version)";
+    homepage = "https://www.ondsel.com";
+    license = licenses.lgpl2Plus;
+    platforms = [ "x86_64-linux" ];
```
Comment (h7x4) [inline pkgs/by-name/on/ondsel/package.nix:34]: > If this package needs a maintainer to be merged, I could do that.

I think that would be a good idea. Are you in the maintainers list?

```suggestion
    platforms = [ "x86_64-linux" ];
    maintainers = with maintainers; [ b-eyselein ];
    mainProgram = "ondsel";
```

Comment (h7x4) [inline pkgs/by-name/on/ondsel/package.nix:34]: I suppose these are supported as well, even though they haven't been packaged. Feel free to mark darwin as broken if ofborg complains and you don't want to deal with it.

```suggestion
    platforms = [ "x86_64-linux" "x86_64-linux" "aarch64-darwin" ];
```

File: pkgs/by-name/on/ondsel/package.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+in
+appimageTools.wrapType2 rec {
+  inherit pname version src;
+
+  extraInstallCommands =
+    let contents = appimageTools.extractType2 { inherit pname version src; };
+    in ''
+      mv $out/bin/${pname}-${version} $out/bin/${pname}
+
+      install -m 444 -D ${contents}/com.ondsel.ES.desktop $out/share/applications/${pname}.desktop
```
Comment (h7x4) [inline pkgs/by-name/on/ondsel/package.nix:22]: nit:

```suggestion
      install -Dm444 ${contents}/com.ondsel.ES.desktop $out/share/applications/${pname}.desktop
```

File: pkgs/by-name/on/ondsel/package.nix (hunk 1 above)
Comment (b-eyselein) [reply pkgs/by-name/on/ondsel/package.nix:34]: I'm not yet in the maintainers list, but I will make a pull request to include myself.

Comment (h7x4) [reply pkgs/by-name/on/ondsel/package.nix:32]: Instead of making a new pull request, could you add yourself in a commit above the existing one?

Comment (b-eyselein) [reply pkgs/by-name/on/ondsel/package.nix:32]: Yes, I can

Comment (b-eyselein) [reply pkgs/by-name/on/ondsel/package.nix:32]: Don't you mean ARM Linux with the second "x86_64-linux" platform?

I could also package it for ARM Linux, but probably not (yet) for MacOS, but I don't know how. Is it something like

```nix
src = if stdenv.isX86() fetchUrl { 
} else if ... fetchUrl {
};
```?

Comment (b-eyselein) [reply pkgs/by-name/on/ondsel/package.nix]: Nevermind, I found another package that packages for different platforms. I've prepared the package for other platforms, but marked all others (ARM Linux, ARM + x86_64 Darwin) as broken.

File: pkgs/by-name/on/ondsel/package.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  suffix = {
+    x86_64-linux = "Linux-x86_64.AppImage";
+    # aarch64-linux = "Linux-aarch64.AppImage";
+    # x86_64-darwin = "macOS-intel-x86_64.dmg";
+    # aarch64-darwin = "macOS-apple-silicon-arm64.dmg";
+  }.${system} or throwSystem;
+
+  src = fetchurl {
+    # TODO: integrate version rest...
```
Comment (h7x4) [inline pkgs/by-name/on/ondsel/package.nix:19]: This todo message isn't very clear. Maybe you could elaborate?

**Comment by h7x4** [comment]
Would you mind squashing the commits down into two?

```
maintainers: add b-eyselein
ondsel: init at 2024.1.0
```

File: pkgs/by-name/on/ondsel/package.nix (hunk 3 above)
Comment (b-eyselein) [reply pkgs/by-name/on/ondsel/package.nix:19]: The current build has a `patch modifier` (maybe some kind of timestamp?) in the url besides the official `major-minor-patch` version (2024.1.0) that I think is going to change in the future. I'm not quite sure if I should integrate it into the url or keep it as a separate variable as it is now. Therefore this TODO

**Comment by b-eyselein** [comment]
I've squashed the commits and deleted the TODO.

**Comment by raspher** [comment]
closes #284524

**Review by pinpox — APPROVED** [review]

**Comment by pinpox** [comment]
Result of `nixpkgs-review pr 284996` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ondsel</li>
  </ul>
</details>

**Review by pinpox — APPROVED** [review]

**Comment by Hayajiro** [comment]
Running nixos-unstable on `x86_64-linux` , the package doesn't seem to work for me. Trying to launch ondsel with varying errors, most of the time the following:
```
❯ ondsel
Ondsel 2024.1.0, Libs: 2024.1.0R35694 (Git)
© FreeCAD team and Ondsel Inc., 2001-2024
Ondsel ES is based on the FreeCAD project and released under the terms of the LGPL2.1+ License.
Ondsel ES wouldn't be possible without the FreeCAD community.
      ###          #           #    ##### #####
     #   #         #           #    #     #
     #   # ###   ### #### #### #    #     #
     #   # #  # #  # #    #    #    ####  #####
     #   # #  # #  # #### ###  #    #         #
     #   # #  # #  #    # #    #    #         #
      ###  #  # #### #### #### ##   ##### #####

free(): invalid pointer
/nix/store/idbjbb813r5i0wyfr2vsrd2ak8zvy51f-ondsel-2024.1.0-extracted/AppRun: line 43: 103384 Aborted                 (core dumped) "${MAIN}" "$@"
```

Happens on multiple systems (combinations of AMD / Intel CPUs and AMD / NVidia GPUs), as well as both using Wayland and X11.

Will look into the details later, but seems like some additional work is required here.

**Comment by pinpox** [comment]
@Hayajiro after updating my system, I'm getting the same error. It was working with nixpkgs form a few days or weeks ago. I was using my flake before which worked fine aswell but is now also getting this error. Maybe it can be useful to narrow down the problem

`nix run 'github:pinpox/ondsel-nix'

**Comment by pinpox** [comment]
Doing some further investigation with `appimage-run` and the appimage used in the package, it seems to be a problem with `LD_LIBRARY_PATH`.

I can run ondsel with no problems by calling the binary directly instead of the generated `AppRun` script. To further investigate this can be reproduced by setting ` export APPIMAGE_DEBUG_EXEC=bash` and then run `appimage-run Ondsel_ES_2024.1.0.35694-Linux-x86_64.AppImage`

**Comment by pinpox** [comment]
@Hayajiro @b-eyselein Have you had any luck with this? I'm kinda dependant on this software but havn't been able to get it to run again. Is there anything I can do or help with to get this fixed?

Let me know how I can help.

**Comment by Hayajiro** [comment]
Sorry, I wasn't able to get to it so far, been busier than usual lately. I'll see if I can get to it in the next couple days. Maybe it would be reasonable to build Ondsel ourselves (just like FreeCAD itself), instead of using the AppImage distribution?

**Comment by pinpox** [comment]
> Maybe it would be reasonable to build Ondsel ourselves (just like FreeCAD itself), instead of using the AppImage distribution?

I have tried to do so here: https://github.com/NixOS/nixpkgs/pull/292995

Fixed a few compilation problems but I'm stuck now at the install step. Maybe you can take a look when you find the time, I'm a bit out of ideas on how to fix it.

**Comment by pinpox** [comment]
@Hayajiro @b-eyselein @raspher I have an alternative PR that builds from source (see above) and works.

**Comment by b-eyselein** [comment]
I've not tested Ondsel with an (recently) updated system, since the application became unresponsive after a settings change, but I've also thought about building Ondsel from source.

Since your PR already does that, I will probably close mine. There is only one small detail I saw: the check `Check pkgs/by-name / check` wants that the package is defined in `pkgs/by-name/on/ondsel`, not under `pkgs/applications/graphics/freecad`...

**Comment by pinpox** [comment]
> the check `Check pkgs/by-name / check` wants that the package is defined in `pkgs/by-name/on/ondsel`

Yes, but I'm not sure that is correct in this case. The package is a fork of freecad, which overrides the source and otherwise uses most of it's derivation. I put it next to freecad the same way other forks (e.g. `orca-slicer` from `bambu-studio`) where packaged. If maintainers really want it in `by-name` I can move it.

There is one remaning issue with my PR though: It builds fine, but even though I specified the ondsel source when I run `result/bin/frecead` it starts freecad instead of ondsel. I'm a bit stuck there, not sure why it does that since it is pulling and building the correct source. The ondsel executable is called "freecad" in the appimage aswell, but displays the ONDSEL ascii art when starting in a terminal. The result from my build still shows the FREECAD ascii.

Maybe someone can give me some pointers

**Comment by Hayajiro** [comment]
I've had a look and the crash looks like it's related to Mesa being too recent. After debugging it myself, I've also found https://github.com/Ondsel-Development/FreeCAD/issues/25 which also hints at Mesa.

**Comment by pinpox** [comment]
> I've had a look and the crash looks like it's related to Mesa being too recent. After debugging it myself, I've also found [Ondsel-Development/FreeCAD#25](https://github.com/Ondsel-Development/FreeCAD/issues/25) which also hints at Mesa.

Could we use an older mesa for the package? How would that be passed to `appimageTools.wrapType2` ?

**Comment by b-eyselein** [comment]
There is an option `extraPkgs` for `appimageTools.wrapType2` that is probably what we've got to use. But the default mesa version on `nixos-stable` (23.11) is `23.1.9`, while some comments on the issue claim that `23.2.1-2` works...

As far as I can tell we'd also have to build mesa ourselves to get an older version, which seems a bit excessive...

**Comment by luzpaz** [comment]
Any progress here ?

**Comment by b-eyselein** [comment]
Sorry, this issue is stale since I'm not using Ondsel anymore. I will close this issue. Maybe the other try #292995 is more successful.

**Pull request closed without merging**